Add the popup widget

Elv13 opened the pull request:
This is the popup widget, a wibox that handles most resizing and placement without
boilerplate code.

This was originally designed to be the final part of the "geometry overhaul" PR
series in 2015-2016, but was never finished.

I rewrote it a few time since then and I am finally less or more happy with the
API. As @psychon would have said, the first few revisions were sledgehammer APIs
that had too many options and did too many things. Thankfully some of these
things, like the shape, are already merged and no longer part of the popup class.
I also dropped a lot of dark magic to automatically move the popup when its parent
moves. It added too much logic for too little gain.

Half of the popup was already merged as `awful.placement.next_to`. The code still
heavily rely on it. I fixed the border_width issue in `next_to` that went
unnoticed since both the original popup (as used in Radical) and the tooltip used
widget  borders.

There is also some more work to make the shims good enough to emulate most of Awesome. It has a a few issues. For some reasons `:find_widgets(x, y)` doesn't work. I didn't investigate much, but it's not so simple to fix (and no, adding more "refresh" doesn't help) and while it messes with some popup examples a bit, I think it can be left for later. This PR is already a monster as-is.

![screenshot_20180113_040418](https://user-images.githubusercontent.com/340384/34904592-d48290de-f816-11e7-9d6a-fb87193c5e31.png)


Finally, the PR also adds the layoutlist widget as it makes a good example use
case. I had this one for ages (all the way back to Awesome 3.3). As the
awful.widget.common template is now merged, it makes more sense to upstream it.

NOTE THAT IT PULLS #2123 commits until they are merged to avoid the GDK warning
on Travis.

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
-## Placing widgets
+### The different type of widget boxes (Wibox)
+
+Awesome API the word "wibox" to describe an area of the screen filled with
```
This sentence... reads weirdly, dunno. Should it be `The Awesome API`, `In the` or something like that? `uses the word`?

How about just mentioning that `wibox` means `widget box`, so a box that contains widgets?

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+### The different type of widget boxes (Wibox)
+
+Awesome API the word "wibox" to describe an area of the screen filled with
+widgets. There is many subvariants of wiboxes with specialized role such as
```
There _are_ many

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+and prevent clients from using this area when tiled.
+
+The `awful.popup` allows to easily place widgets on the screen. It automatically
+resize itself to fit the optimal widget size. It also have helper properties
```
have -> has

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+
+The `awful.popup` allows to easily place widgets on the screen. It automatically
+resize itself to fit the optimal widget size. It also have helper properties
+and method to make it easy to place it on the screen. It both absolute
```
method -> methods

psychon replied:
it both _supports_ absolute

(Also, I wouldn't use `both` on a list with three entries, how about `It supports all of` or just `It supports absolute, relative and manual positioning`?)

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+The `awful.tooltip` is a very simple `wibox` that allows to display text next
+to an object such as the mouse.
+
+Finally, the `awful.titlebar`, while not technically a real `wibox`, act
```
act_s_

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+to an object such as the mouse.
+
+Finally, the `awful.titlebar`, while not technically a real `wibox`, act
+exactly the same way and allows to attach widgets on each side of the clients.
```
of clients

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+
+## The different syntaxes to initiate widgets
+
+Awesome provides 2 totally different API access style to manage widgets. Both
```
style_s_

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+## The different syntaxes to initiate widgets
+
+Awesome provides 2 totally different API access style to manage widgets. Both
+suit different use case and are execute exactly the same code under the hood.
```
What does "are execute" mean? `and are based on the same code`?

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
 
 @DOC_layout_WIDGET_LIST@
 
-## Placing widgets
```
If you want, I can produce a fixup commit for the following comments, although some of them are questions instead of just plain suggestions.

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+suit different use case and are execute exactly the same code under the hood.
+
+The **imperative** widget initialization is similar to QtWidgets, GTK and Win32.
+You create the object, the set the property and add the widget as a child to
```
the -> then
(Also `the property` "feels" weird to me, `set properties`?)

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+The **imperative** widget initialization is similar to QtWidgets, GTK and Win32.
+You create the object, the set the property and add the widget as a child to
+another already declared widget. It is quite simple to use but very verbose
+and full of boilerplate code. The imperative API also offers property both with
```
propert_ies_

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+widgets are added and removed over the course of their lifecycle.
+
+The **declarative** syntax resembles HTML style code
+written in JSON or YAML. The widgets instances are created automatically and
```
HTML written in JSON or YAML?!?!?!?!?!?!?!?!?!?

psychon replied:
Okay:

The **declarative** syntax resembles HTML-style code in that an explicit nesting is present and is used to express "boxes containing boxes" (the box model). However, instead of requiring (heavy-weight) XML, a lighter JSON-like syntax is used.

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+
+The **declarative** syntax resembles HTML style code
+written in JSON or YAML. The widgets instances are created automatically and
+the hierarchy is related to the indentation. It is preferred when creating
```
I can also use the imperative system together with "funny" indentation. :-)
How about `table nesting` instead? I'm nut quite sure...

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+The **declarative** syntax resembles HTML style code
+written in JSON or YAML. The widgets instances are created automatically and
+the hierarchy is related to the indentation. It is preferred when creating
+static layouts that wont change over the course of their lifecycle.
```
won't

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+    bg:set_bg("#ff0000")
+
+    local tb1 = wibox.widget.textbox("foo")
+    local tb2 = wibox.widget.textbox("bar")
```
How about using `tb1:set_text("foo")` instead, just so that there are more differences between the styles?

psychon commented on docs/03-declarative-layout.md:
```diff
@@ -29,7 +28,116 @@ configurable rules.
+but both are **always** available. Note that each style can be mixed with other
+styles, but this creates very confusing code and should be avoided.
+
+## Creating and placing widgets using the declarative
```
using the declarative what? style?

psychon commented on tests/examples/awful/popup/wiboxtypes.lua:
```diff
@@ -0,0 +1,237 @@
+}
+
+-- poor copy of awful.widget.calendar_widget until I fix its API to be less
+-- insane.
```
I love how everyone loves this :-)

psychon commented on lib/awful/widget/layoutlist.lua:
```diff
@@ -0,0 +1,151 @@
+
+    local wdg = wibox.widget.base.make_widget(nil, nil, {
+        enable_properties = true,
+    })
```
Shouldn't this set `_private.layout` somewhere? The only place setting that seems to be `set_base_layout`?!?

Elv13 replied:
done

psychon commented on lib/awful/widget/layoutlist.lua:
```diff
@@ -0,0 +1,151 @@
+    wdg._private.data = {}
+    gtable.crush(wdg, args)
+
+    tag.connect_signal("property::layout", function() reload_cache(wdg); update(wdg) end)
```
Please wire up this widget to `test-leaks.lua` and then make the widget somehow GC'able.

psychon commented on lib/awful/widget/layoutlist.lua:
```diff
@@ -0,0 +1,151 @@
+        default_buttons,
+        wb_label,
+        self._private.data,
+        self._private.cache,
```
Why is the cache necessary here? `gears.surface.load` has its own cache and besides that `reload_cache` doesn't do anything expensive. Just use `alayout.layouts` directly here and do "the necessary stuff" in `wb_label`.

Elv13 replied:
The other `awful.widget.common` widgets work like that. I could be changed later on.

psychon commented on lib/awful/widget/layoutlist.lua:
```diff
@@ -0,0 +1,151 @@
+
+local default_buttons = gtable.join(
+    abutton({ }, 1, function(a) a.callback() end)
+)
```
Correct me if I'm wrong, but this is not the default, but the hardcoded list of buttons that cannot be overridden. Plus, what does this button do anyway? `callback` is always an empty function?!?

Elv13 replied:
ah, indeed, I forgot to actually implement it, done

Elv13 replied:
done

psychon commented on lib/awful/widget/layoutlist.lua:
```diff
@@ -0,0 +1,151 @@
+
+local function wb_label(item)
+    local selected = alayout.get(item.screen) == item.layout
+    return item.layout.name, selected and beautiful.bg_focus or nil, nil, item.icon
```
Shouldn't there be a `beautiful.bg_layoutlist` in here?

Elv13 replied:
done (doubling the file size... whatever)

psychon commented on lib/awful/widget/layoutlist.lua:
```diff
@@ -0,0 +1,151 @@
+-- @tparam widget args.layout The widget layout (not to be confused with client
+--  layout).
+-- @tparam[opt] table args.widget_template A custom widget to be used for each action.
+-- @tparam screen args.screen A screen (mandatory)
```
Some `assert` should check for this mandatory argument, I think

Elv13 replied:
done

psychon commented on tests/examples/awful/widget/layoutlist/bar.lua:
```diff
@@ -0,0 +1,62 @@
+local awful = { --DOC_HIDE --DOC_NO_USAGE
```
Could you please also do a test showing the "default behaviour" of the layoutlist, i.e. only the minimum required arguments are given (`screen = 1`, I think). That way, (a) I'm more convinced that the widget actually works by default and (b) people get an idea what the widget actually looks like by default.

Elv13 replied:
done

psychon commented on tests/examples/shims/beautiful.lua:
```diff
@@ -34,6 +34,8 @@ local module = {
     border_width = 1.5        ,
 
+    prompt_bg_cursor = "#00FF7D",
+
```
Huh? Got some details for what this commit fixes exactly? For example, an error message with traceback? How is this related with 8cc469322f8043b262b36dd7743bb330bdde753d?

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+
+    local preferred_anchors = self._private.preferred_anchors or {
+        "back", "front", "middle"
+    }
```
done

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+
+    if anchor_name ~= self._private.current_anchor then
+        self:emit_signal("property::current_anchor", anchor_name)
+        self._private.current_anchor = anchor_name
```
whoops, copy paste gone wrong, done

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+-- Layout this widget
+function main_widget:layout(context, width, height)
+    if self.widget then
+        local w, h = wibox.widget.base.fit_widget(self, context, self.widget, 9999, 9999)
```
> but I feel like they should do so way "harder"

What do you have in mind? I don't feel like adding memes to the doc ;)

> How about using context.screen.geometry.width/height here instead?

I could add a new `awful.screen` function to get the maximum width/height of the screens. I can't just use the current screen because it can change in some async events.

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+            if self._wb.width ~= prev_geo.width and self._wb.height ~= prev_geo.height then
+                set_position(self._wb)
+            end
+        end)
```
> Why does it need to reach for idle_add instead of gears.timer.delayed_call?

The code is very old, I updated it.

> Why does it need to be delayed at all and cannot be done immediately?

An image is worth a thousand word (and you might actually not believe me)

![screenshot_20180115_021433](https://user-images.githubusercontent.com/340384/34930889-d2ae9616-f999-11e7-83bd-baff51eaee4e.png)

My theory is that the xcb surface is not reallocated yet

psychon replied:
Uhm... Do you happen to know an easy way to reproduce this? At least the following does not work.
```lua
gears.timer.start_new(2, function()
	local w = wibox{ x = 50, y = 50, width = 100, height = 10, bg = "#ff0000", fg = "#000000" }
	w:set_widget(wibox.widget.textbox("Foo"))
	w.visible = true
	gears.timer.delayed_call(function()
		w.height = 50
	end)
end)
```

Elv13 replied:
If you apply the patches, it is quite easy to reproduce by commenting the `delayed_call` line. I am sorry for not having a smaller test case. I know about this issue for a years, but I always considered it irrelevant given the kind of hack you have to use to cause it (the popup is such an hack, Radical was another).

psychon replied:
Ah. Stderr says (however, I did not get visual corruption, only this)
```
2018-01-15 10:03:12 E: awesome: Error during a protected call: lib/wibox/drawable.lua:165: Cairo context entered error state: SURFACE_FINISHED
stack traceback:
	[C]: in function 'assert'
	lib/wibox/drawable.lua:165: in upvalue 'do_redraw'
	lib/wibox/drawable.lua:390: in function <lib/wibox/drawable.lua:388>
	[C]: in function 'xpcall'
	lib/gears/protected_call.lua:36: in function <lib/gears/protected_call.lua:35>
	(...tail calls...)
	lib/gears/timer.lua:231: in function <lib/gears/timer.lua:229>
```
This is the opposite of your theory: The problem is that the xcb surface is reallocated while Lua code still wants to draw to it. Thus, at the end after all the drawing it noticed that something is wrong. (The above happens just by doing `awful.popup{ widget = wibox.widget.textbox("textbox") }`).

So, basically: You may not resize the wibox that is currently drawn while it is being drawn.

...which brings us back to this ugly delayed call and all of its problems.

How much would you be against the popup creating its own `wibox.hierarchy` instance and using that for catching relayouts instead of this magic in `:layout`? That way, this would be disconnected from the widget drawing code instead of being a side effect of it. (And then both your `:fit` and `:layout` are boilerplate and using the argument to `wibox.widget.base.make_widget` makes even more sense... well, no, your whole `main_widget` becomes unnecessary and instead you can just directly give the widget to the wibox)

I wanted to hack this idea into your code, but got lost, so instead here is it in self-contained:
```lua
local widget = wibox.widget.textbox("textbox")
-- Resize the textbox in 5 seconds
gears.timer.start_new(5, function() widget:set_text("This is a longer text than what is was before") end)

local set_widget
do
        local wrapper = wibox.widget.base.make_widget() -- This only needs :layout(), other functions are never used
        local widget = widget
        function wrapper:layout(context, width, height)
                return { wibox.widget.base.place_widget_at(widget, 0, 0,
                        wibox.widget.base.fit_widget(self, context, widget, width, height)) }
        end
        local w = wibox{ x = 50, y = 50, visible = true, widget = widget } -- This displays widget directly, not wrapper!
        local ctxt = { meh_need_to_fake_a_context = true, dpi = screen.primary.dpi }
        local h
        local function layout_callback()
            h:update(ctxt, wrapper, 2^16-1, 2^16-1)
            print(h:get_children()[1]:get_size())
            w.width, w.height = h:get_children()[1]:get_size()
        end
        h = wibox.hierarchy.new(ctxt, wrapper, 2^16-1, 2^16-1, function() end, layout_callback, {})
        layout_callback()
        function set_widget(arg)
                w.widget = arg
                widget = arg
                wrapper:emit_signal("widget::layout_changed")
        end
end

gears.timer.start_new(10, function() set_widget(wibox.widget.textbox("Show a different widget")) end)
```

psychon replied:
Or alternatively with `:layout` resizing the wibox again directly:
```lua
local widget = wibox.widget.textbox("textbox")
-- Resize the textbox in 5 seconds
gears.timer.start_new(5, function() widget:set_text("This is a longer text than what is was before") end)

local set_widget
do
        local wrapper = wibox.widget.base.make_widget() -- This only needs :layout(), other functions are never used
        local widget = widget
        local w = wibox{ x = 50, y = 50, visible = true, widget = widget } -- This displays widget directly, not wrapper!
        function wrapper:layout(context, width, height)
                local wi, h = wibox.widget.base.fit_widget(self, context, widget, width, height)
                w.width = wi
                w.height = h
                return { wibox.widget.base.place_widget_at(widget, 0, 0, wi, h) }
        end
        local ctxt = { meh_need_to_fake_a_context = true, dpi = screen.primary.dpi }
        local h
        local function layout_callback()
            h:update(ctxt, wrapper, 2^16-1, 2^16-1)
        end
        h = wibox.hierarchy.new(ctxt, wrapper, 0, 0, function() end, layout_callback, {})
        layout_callback()
        function set_widget(arg)
                w.widget = arg
                widget = arg
                wrapper:emit_signal("widget::layout_changed")
        end
end

gears.timer.start_new(10, function() set_widget(wibox.widget.textbox("Show a different widget")) end)
```

Elv13 replied:
Umm, I kind of like the idea, but my limited attempts created a result more problematic than the current version. I guess "I got lost" the same way you did when you tried to implement this.

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+function main_widget:set_children(children)
+    self.widget = children and children[1]
+    self:emit_signal("widget::layout_changed")
+end
```
I think it would add some boilerplate of it's own. There's only the `:fit()` boilerplate. I don't see this as an issue.

psychon replied:
Well, that and `set_children`, `get_children`. Even if it just saves 10 lines, that's 10 lines less code. ;-)

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+
+    self._private.widget_geo = obj
+
+    if not obj then return end
```
done

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+    if not obj then return end
+
+    if obj._apply_size_now then
+        obj:_apply_size_now(false)
```
It is necessary because popups can be placed in relation to each other (like the notifications) and at some point it is necessary to collapse all the async stuff. But the async stuff are also necessary for some other corner case like the screenshot above. Without the async part, it can actually crash Awesome.

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+
+    local oldoff = self._private.offset or {}
+    local oldx   = oldoff.x or 0
+    local oldy   = oldoff.y or 0
```
Because X could be set and not Y? Anyway, I refactored it so it cannot happen. It renders the comment below invalid.

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+end
+
+-- For the tests and the race condition when 2 popups are placed next to each
+-- other.
```
It will crash

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+end
+
+--- A brilliant idea to totally turn the whole hierarchy on its head
+-- and create a widget that own a wibox...
```
Whoops, leftover from the prototype (it was very, very hacky back then, way more than the still rather hacky code)

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+
+--@DOC_wibox_COMMON@
+
+return setmetatable(module, {__call = create_auto_resize_widget})
```
done

Elv13 commented:
I added a couple commits to fix some of the `popup` issues reported by @psychon. I also added `minimum` and `maximum` properties to better address the "this will explode if the widget size diverge" issue.

Elv13 commented:
Note that I added a new commit https://github.com/awesomeWM/awesome/pull/2154/commits/3a58777b8066d70a33464b65eef9e072f5b8fbee to try to fix the build on Lua 5.1. But I still can't reproduce and it doesn't fix it. Anyhow, not using `next()` and `pairs()` is probably a good thing in itself so I will leave it there.

Elv13 commented:
Rebased to remove the duplicate commits from the titlebar doc PR (as they were required to pass the test). Otherwise there is still the `awful.popup` way to updating it's size that remains to be done. There is no really valuable way to split this as all tests of other components require the popup class.

Elv13 commented on tests/examples/awful/template.lua:
```diff
@@ -164,43 +187,45 @@ end
-                client_widget(c, c.color or geo._color or beautiful.bg_normal, geo._label),
-                {x=geo.x, y=geo.y}
-            )
+    if not c.minimized then
```
done (sorry for all the "done" comments I will add, it just hard to keep track of what's done in GitHub given how arbitrary it's "outdated comment" feature is)

psychon replied:
I don't mind "done" comments. In fact, I think they are a really good idea and should be used more often. :-)

Elv13 commented on tests/examples/shims/client.lua:
```diff
@@ -120,6 +120,20 @@ function client.gen_fake(args)
+    ret.drawin = ret
+    ret.drawable = ret
+
+    -- Make sure the layer properties are not `nil`
```
Everything that use `if foo == false then ... end` code doesn't work, as simple as that

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+
+function popup:set_widget(wid)
+    self._private.widget = wid
+    self._private.container:set_widget(wid)
```
done

Elv13 commented on lib/awful/popup.lua:
```diff
@@ -0,0 +1,434 @@
+
+    -- Temporarily remove the widget
+    local original_widget = args.widget
+    args.widget = nil
```
done

Elv13 commented on lib/wibox/init.lua:
```diff
@@ -253,6 +253,10 @@ local function new(args)
 
+    function ret._drawable.get_wibox()
+        return ret
+    end
```
> (And I fail to see how this commit has anything to do with find_widgets even though the commit 
> summary says that the commit is all about that)

Because the clickable or mouse_overable widgets callback information comes from `find_widgets` and if that's used to move a wibox then it has to know the wibox. I agree it's a bit hacky, but fixing that would be a bit complex unless I missed something obvious when I wrote that (didn't look recently). Feel free to propose a better solution, but I don't have much AwesomeWM time and a backlog of > 500 commits to finish, so it's a bit hopeless to expect me working on this in any kind of timely fashion.

Elv13 commented:
@psychon Any idea about the race condition that causes the popup to move to the wrong side (the one that makes the test fail)?

I added some `print()` in some older revision of the PR and it seems random. However since it doesn't happen without this PR, I guess it has to be related to the `awful.placement.next_to` changes. It happens in some commit, not others and not all the same ones each time. It also only seem to happen on luajit since I 1) can't reproduce at all 2) had this PR run 35+ times on Travis and never saw the issue in the other builders.

Also, I think I am mostly done with the other changes. A fresh review would be welcome. I didn't change much since 3 months, it's mostly a rebase with minor documentation comment changes.

psychon commented:
> @psychon Any idea about the race condition that causes the popup to move to the wrong side (the one that makes the test fail)?

On Travis, only the LuaJIT build failed. And in fact, it failed because of `TEST_PREV_COMMITS=1`. And here things start to go weird. The summary at the end claims that the tests failed for commit aaa0d92, but when looking at the attempted build with that commit, it succeeded. Oh, great, a bug in the output of this thing... #2261 fixes that (hopefully, I cannot really test it).
Uhm, and from a quick look, this PR does not even touch that test...? Sorry, but debugging `awful.placement` is beyond my abilities.

psychon commented on tests/examples/awful/template.lua:
```diff
@@ -154,6 +159,24 @@ local function client_widget(c, col, label)
+    local w = d.get_wibox and d:get_wibox() or nil
+    if w then
+        -- Force a full layout first as widgets with as the awful.popup have
+        -- interdependencies between the content and the container
```
I still fail to parse this comment as English.
- "Force a full layout first": Okay, so before anything else is done, something(?) is (re-?)layouted first.
- "as widgets": Because of widgets
- "with as the": Sorry, I'm lost

This is the first commit that I look at and just the second hunk in the diff...

psychon commented:
Okay... so, I built lgi against LuaJIT:
```diff
diff --git a/lgi/Makefile b/lgi/Makefile
index 08419bd..e1da776 100644
--- a/lgi/Makefile
+++ b/lgi/Makefile
@@ -43,6 +43,7 @@ endif
 ifeq ($(HOST_OS),darwin)
 CFLAGS += -DGOBJECT_INTROSPECTION_LIBDIR=\"$(GOBJECT_INTROSPECTION_LIBDIR)\"
 endif
+LUA_CFLAGS = -I /usr/include/luajit-2.1/
 ALL_CFLAGS = $(CCSHARED) $(COPTFLAGS) $(LUA_CFLAGS) $(shell $(PKG_CONFIG) --cflags $(PKGS)) $(CFLAGS)
 LIBS += $(shell $(PKG_CONFIG) --libs $(PKGS))
 ALL_LDFLAGS = $(LIBFLAG) $(LDFLAGS)
```
Then I built awesome against LuaJIT (we might want to simplify this...):
```diff
diff --git a/awesomeConfig.cmake b/awesomeConfig.cmake
index 892b88e6..454c272e 100644
--- a/awesomeConfig.cmake
+++ b/awesomeConfig.cmake
@@ -358,7 +358,7 @@ file(MAKE_DIRECTORY ${BUILD_DIR}/script_files/)
 
 add_custom_command(
         OUTPUT ${BUILD_DIR}/docs/06-appearance.md
-        COMMAND lua ${SOURCE_DIR}/docs/06-appearance.md.lua
+        COMMAND luajit ${SOURCE_DIR}/docs/06-appearance.md.lua
         ${BUILD_DIR}/docs/06-appearance.md
         DEPENDS lgi-check-run
 )
diff --git a/tests/examples/CMakeLists.txt b/tests/examples/CMakeLists.txt
index 124e5b1b..88ab9810 100644
--- a/tests/examples/CMakeLists.txt
+++ b/tests/examples/CMakeLists.txt
@@ -17,7 +17,7 @@ endif()
 cmake_minimum_required(VERSION 3.0.0)
 
 # Get and update the LUA_PATH so the scripts can be executed without awesome.
-execute_process(COMMAND lua -e "p = package.path:gsub(';', '\\\\;'); io.stdout:write(p)"
+execute_process(COMMAND luajit -e "p = package.path:gsub(';', '\\\\;'); io.stdout:write(p)"
     OUTPUT_VARIABLE "LUA_PATH_")
 
 # Make sure the system can be called from the test directory
@@ -32,7 +32,7 @@ endif()
 
 if (DO_COVERAGE)
   execute_process(
-    COMMAND lua -e "require('luacov.runner')('${TOP_SOURCE_DIR}/.luacov')"
+    COMMAND luajit -e "require('luacov.runner')('${TOP_SOURCE_DIR}/.luacov')"
     RESULT_VARIABLE TEST_RESULT
     ERROR_VARIABLE  TEST_ERROR
     ERROR_STRIP_TRAILING_WHITESPACE)
@@ -40,9 +40,9 @@ if (DO_COVERAGE)
     message(${TEST_ERROR})
     message(FATAL_ERROR "Failed to run luacov.runner.")
   endif()
-  set(LUA_COV_RUNNER lua "-erequire('luacov.runner')('${TOP_SOURCE_DIR}/.luacov')")
+  set(LUA_COV_RUNNER luajit "-erequire('luacov.runner')('${TOP_SOURCE_DIR}/.luacov')")
 else()
-  set(LUA_COV_RUNNER lua)
+  set(LUA_COV_RUNNER luajit)
 endif()
 
 # Add the main awesome lua libraries.
```
Plus: `LUA_CPATH="/tmp/lgi/?.so;;" LUA_PATH="/tmp/lgi/?.lua;/tmp/lgi/?/init.lua;;" make CMAKE_ARGS="-DLUA_LIBRARY=/usr/lib64/libluajit-5.1.so -DLUA_INCLUDE_DIR=/usr/include/luajit-2.1"`

According to Travis, the following should fail now:
```
git checkout 84c2c961f22bc48423874818f3a64568f66be092
LUA_CPATH="/tmp/lgi/?.so;;" LUA_PATH="/tmp/lgi/?.lua;/tmp/lgi/?/init.lua;;"  tests/run.sh /tmp/awesome/tests/test-tooltip.lua
```
And... well, it does fail. So now comes the debugging fun, which takes forever thanks to having to run `make`, which rebuilds all the docs all the time.

Okay, so here is a patch that adds lots of debugging output:
```diff
diff --git a/lib/awful/placement.lua b/lib/awful/placement.lua
index 4499c5eb..7fd9c72b 100644
--- a/lib/awful/placement.lua
+++ b/lib/awful/placement.lua
@@ -1386,24 +1386,46 @@ function placement.next_to(d, args)
 
     local regions = get_relative_regions(wgeo, mode, is_absolute)
 
+    require("gears.debug").dump({
+        input_geo = wgeo,
+        is_absolute = is_absolute,
+        regions = regions
+    }, nil, 2)
+    require("gears.debug").dump(regions)
+
     -- Check each possible slot around the drawable (8 total), see what fits
     -- and order them by preferred_positions
     local does_fit = {}
     for k,v in pairs(regions) do
+        print("Checking", k, v)
         local geo, dir = outer_positions[k.."1"](v, dgeo.width, dgeo.height)
         geo.width, geo.height = dgeo.width, dgeo.height
         local fit = fit_in_bounding(v.screen, geo, args)
 
+        require("gears.debug").dump {
+            geo = geo,
+            dir = dir,
+            fit = fit
+        }
+
         -- Try the other compatible geometry
         if not fit then
+            print("Trying compatible geometry")
             geo, dir = outer_positions[k.."2"](v, dgeo.width, dgeo.height)
             geo.width, geo.height = dgeo.width, dgeo.height
             fit = fit_in_bounding(v.screen, geo, args)
+
+            require("gears.debug").dump {
+                geo = geo,
+                dir = dir,
+                fit = fit
+            }
         end
 
         does_fit[k] = fit and {geo, dir} or nil
 
         if fit and preferred_positions[k] and preferred_positions[k] < pref_idx then
+            print("Preferred!")
             pref_idx  = preferred_positions[k]
             pref_name = k
         end
diff --git a/lib/awful/tooltip.lua b/lib/awful/tooltip.lua
index 36f86933..6b37f9f9 100644
--- a/lib/awful/tooltip.lua
+++ b/lib/awful/tooltip.lua
@@ -181,6 +181,11 @@ local function apply_mouse_mode(self)
     local align          = self._private.align
     local real_placement = align_convert[align]
 
+    require("gears.debug").dump{ mode = "mouse",
+    align = self._private.align,
+    real_placement = align_convert[align]
+}
+
     a_placement[real_placement](w, {
         parent = capi.mouse,
         offset = offset[align]
@@ -190,12 +195,17 @@ end
 local function apply_outside_mode(self)
     local w = self:get_wibox()
 
+    require("gears.debug").dump({ mode = "outside",
+    geometry = self._private.widget_geometry,
+    prefered = self.preffered_positions
+}, nil, 2)
     local _, position = a_placement.next_to(w, {
         geometry            = self._private.widget_geometry,
         preferred_positions = self.preferred_positions,
         honor_workarea      = true,
     })
 
+    print("Chose position:", position)
     if position ~= self.current_position then
         -- Re-apply the shape.
         apply_shape(self)
diff --git a/tests/test-tooltip.lua b/tests/test-tooltip.lua
index d545ec56..e9159cb5 100644
--- a/tests/test-tooltip.lua
+++ b/tests/test-tooltip.lua
@@ -89,17 +89,20 @@ table.insert(steps, function()
 
     tt:add_to_object(imb)
 
+    print("Moving the mouse")
     -- Move the mouse over the widget
     place.top_left(mouse)
     mouse.coords {
         x = w.x + w.width - 20 - 12.5,
         y = w.y + 125 + 12.5,
     }
+    print("Moved mouse to x=", w.x + w.width - 20 - 12.5, " and y=", w.y + 125 + 12.5)
     return true
 end)
 
 -- Test that the above move had the intended effect
 table.insert(steps, function()
+    print("Checking current position is top")
     assert(tt.current_position == "top", tt.current_position)
 
     return true
```
The relevant part of the output is:
```
Moving the mouse
Moved mouse to x=	604.5	 and y=	396.5
table: 0x41d8ad60
  geometry : table: 0x41d8a5b0
    x : 205 (number)
    widget : @lib.wibox.widget.imagebox (table: 0x41f8da10) (table) […]
    y : 125 (number)
    widget_height : 25 (number)
    hierarchy : table: 0x41f2a0a0 (table) […]
    widget_width : 25 (number)
    width : 25 (number)
    drawable : wibox drawable (@.tmp.awesome.tests.test-tooltip) (table: 0x405938f0) (table) […]
    height : 25 (number)
  mode : outside (string)
table: 0x41d8cb30
  input_geo : table: 0x41d8a5b0
    x : 205 (number)
    widget : @lib.wibox.widget.imagebox (table: 0x41f8da10) (table) […]
    y : 125 (number)
    widget_height : 25 (number)
    hierarchy : table: 0x41f2a0a0 (table) […]
    widget_width : 25 (number)
    width : 25 (number)
    drawable : wibox drawable (@.tmp.awesome.tests.test-tooltip) (table: 0x405938f0) (table) […]
    height : 25 (number)
  is_absolute : false (boolean)
  regions : table: 0x41d8c9b8
    right : table: 0x41d8caa8 (table) […]
    left : table: 0x41d8ca48 (table) […]
    bottom : table: 0x41d8cb68 (table) […]
    top : table: 0x41d8cb08 (table) […]
table: 0x41d8c9b8
  right : table: 0x41d8caa8
    x : 1 (number)
    screen : screen: 0x419df8e0 (screen)
    y : 125 (number)
    distance : 216.86055427394 (number)
    height : 25 (number)
    width : 25 (number)
  left : table: 0x41d8ca48
    x : 0 (number)
    screen : screen: 0x419df8e0 (screen)
    y : 125 (number)
    distance : 217.85889929034 (number)
    height : 25 (number)
    width : 25 (number)
  bottom : table: 0x41d8cb68
    x : 205 (number)
    screen : screen: 0x419df8e0 (screen)
    y : 1 (number)
    distance : 137.07114940789 (number)
    height : 25 (number)
    width : 25 (number)
  top : table: 0x41d8cb08
    x : 205 (number)
    screen : screen: 0x419df8e0 (screen)
    y : 0 (number)
    distance : 138.06701271484 (number)
    height : 25 (number)
    width : 25 (number)
Checking	right	table: 0x41d8caa8
table: 0x41d904b0
  geo : table: 0x41d900a0
    y : 125 (number)
    x : 1 (number)
    height : 14 (number)
    width : 62 (number)
  fit : true (boolean)
  dir : down (string)
Checking	left	table: 0x41d8ca48
table: 0x41d90e28
  geo : table: 0x41d90a18
    y : 125 (number)
    x : -62 (number)
    height : 14 (number)
    width : 62 (number)
  fit : false (boolean)
  dir : down (string)
Trying compatible geometry
table: 0x41d91a28
  geo : table: 0x41d913a0
    y : 136 (number)
    x : -62 (number)
    height : 14 (number)
    width : 62 (number)
  fit : false (boolean)
  dir : up (string)
Checking	bottom	table: 0x41d8cb68
table: 0x41d92508
  geo : table: 0x41d920f8
    y : 1 (number)
    x : 205 (number)
    height : 14 (number)
    width : 62 (number)
  fit : false (boolean)
  dir : right (string)
Trying compatible geometry
table: 0x41d930a0
  geo : table: 0x41d92c90
    y : 1 (number)
    x : 168 (number)
    height : 14 (number)
    width : 62 (number)
  fit : false (boolean)
  dir : left (string)
Checking	top	table: 0x41d8cb08
table: 0x41d93be8
  geo : table: 0x41d937d8
    y : -14 (number)
    x : 205 (number)
    height : 14 (number)
    width : 62 (number)
  fit : false (boolean)
  dir : right (string)
Trying compatible geometry
table: 0x41d94540
  geo : table: 0x41d94130
    y : -14 (number)
    x : 168 (number)
    height : 14 (number)
    width : 62 (number)
  fit : false (boolean)
  dir : left (string)
Chose position:	right
Checking current position is top
Error: running function for step 13/16 (@1): /tmp/awesome/tests/test-tooltip.lua:106: right
```
So, top was not chosen because that would lead to `y=-14`, aka partly offscreen.
The same test with Lua 5.3 produces:
Sigh. The same error and pretty much the same output.

@Elv13 Why do you think exactly that this is LuaJIT-specific? How about this being `TEST_PREV_COMMITS=1` specific instead? In fact, I just tried and now `test-tooltip.lua` also fails for the HEAD of this branch with either Lua 5.3 or LuaJIT. But something is definitely wrong, here is a screenshot taken after letting the test run under `TEST_PAUSE_ON_ERRORS=1`:
![foo](https://user-images.githubusercontent.com/89482/43087955-51c001aa-8ea1-11e8-8300-9686d92f2269.png)
That tooltip looks a bit... off?

Elv13 commented:
Rebased on top of the merged commits. This PR has 97% coverage.

As said in the v4.3 roadmap comment in another PR, the big remaining item here is the `delayed_call` instead of the hierarchy. I propose to skip this for now even if I agree it is conceptually better. Integrating this change has so far produced regressions and I would appreciate for this to be in v4.3 to allow the code to be used before I propose the `naughty` pull request that makes heavy use of it. I think it can be made to work without the inconsistencies, so I propose to open an issue and do it for v4.4.

Tracking this over such a large PR is hard but I think most comments were handled earlier this year. The tooltip regression that caused some builds to pass and some to fail is still troubling and the root cause wasn't identified. The bug itself was pretty trivial to fix using git-bisect, but I am not really happy about it. There is a chicken and eggs that causes the tooltip to require this pull request before it can have documentation tests to prevent future heisenbugs. I don't want to add more code to this PR because if its huge size. However it can be added to the v4.3 milestone. It's trivial to write once the template and placement commits are merged. The new improvements you added to the documentation artifacts will allow to detect future changes/regressions.

So, @psychon how do you want to go forward about this. I can split it in chunks but they are interdependent so doing this would force all the tests to go in the last PR. So it would leave some `@DOC_...EXAMPLE@` that point to nowhere at first (and no, I wont create commits to add them in that PR, it makes rebasing a mess). Or you can review the whole thing again and (hopefully) be done with it. It hasn't changed beside being rebased a hundred time, so the old review applies.

psychon commented:
@Elv13 By the way, what happened to the sometimes-broken test? According to the screenshot from my last comment, `awful.placement` is broken. And I think "sometimes randomly broken" is worse than "always broken".

Elv13 commented:
It now exists in smaller PRs
